corehq/apps/users/user_data.py

```diff
@@ -31,27 +31,44 @@ def lazy_init(cls, couch_user, domain):
         raw_user_data = couch_user.to_json().get('user_data', {}).copy()
         raw_user_data.pop(COMMCARE_PROJECT, None)
         profile_id = raw_user_data.pop(PROFILE_SLUG, None)
-        sql_data, _ = SQLUserData.objects.get_or_create(
-            user_id=couch_user.user_id,
-            domain=domain,
-            defaults={
-                'data': raw_user_data,
-                'django_user': couch_user.get_django_user,
-                'profile_id': profile_id,
-            }
-        )
+
+        try:
+            sql_data = SQLUserData.objects.get(user_id=couch_user.user_id, domain=domain)
+        except SQLUserData.DoesNotExist:
+            if not (raw_user_data or profile_id):
+                # Don't bother saving anything to the DB
+                return cls({}, couch_user, domain)
+
+            sql_data = SQLUserData.objects.create(
+                user_id=couch_user.user_id,
+                domain=domain,
+                data=raw_user_data,
+                django_user=couch_user.get_django_user(),
+                profile_id=profile_id,
+            )
+
         return cls(sql_data.data, couch_user, domain, profile_id=sql_data.profile_id)
 
+    @transaction.atomic
     def save(self):
-        SQLUserData.objects.update_or_create(
-            user_id=self._couch_user.user_id,
-            domain=self.domain,
-            defaults={
-                'data': self._local_to_user,
-                'django_user': self._couch_user.get_django_user,
-                'profile_id': self.profile_id,
-            },
-        )
+        try:
+            sql_data = (SQLUserData.objects
+                        .select_for_update()
+                        .get(user_id=self._couch_user.user_id, domain=self.domain))
+        except SQLUserData.DoesNotExist:
+            if self._local_to_user or self.profile_id:
+                SQLUserData.objects.create(
+                    user_id=self._couch_user.user_id,
+                    domain=self.domain,
+                    data=self._local_to_user,
+                    django_user=self._couch_user.get_django_user(),
+                    profile_id=self.profile_id,
+                )
+            # Don't save to DB if there's nothing to save
```

Comment:
this comment is for the else block if I am not mistaken.

Nit. indent it the same depth

Reply:
it's not for the `else` block on the line below this, it's for the condition where the `except` is triggered, but the `if` above this does not pass.  I can rephrase it and put it on the `if`